adapters/httpapi_winhttp.c

```diff
@@ -740,6 +826,65 @@ HTTPAPI_RESULT HTTPAPI_SetOption(HTTP_HANDLE handle, const char* optionName, con
             /*winhttp accepts all certificates, because it actually relies on the system ones*/
             result = HTTPAPI_OK;
         }
+        else if (strcmp(OPTION_HTTP_PROXY, optionName) == 0)
+        {
+            char proxyAddressAndPort[MAX_HOSTNAME_LEN];
+            HTTP_PROXY_OPTIONS* proxy_data = (HTTP_PROXY_OPTIONS*)value;
+            if (proxy_data->host_address == NULL || proxy_data->port <= 0)
+            {
+              LogError("invalid proxy_data values ( host_address = %p, port = %d)", proxy_data->host_address, proxy_data->port);
+              result = HTTPAPI_ERROR;
+            }
+            else
+            {
+                if (sprintf_s(proxyAddressAndPort, MAX_HOSTNAME_LEN, "%s:%d", proxy_data->host_address, proxy_data->port) <= 0)
+                {
+                    LogError("failure constructing proxy address");
+                    result = HTTPAPI_ERROR;
+                }
+                else
+                {
+                    if(mallocAndStrcpy_s((char**)&(httpHandleData->proxy_host), (const char*)proxyAddressAndPort) != 0)
+                    {
+                        LogError("failure allocate proxy host");
+                        result = HTTPAPI_ERROR;
+                    }
+                    else
+                    {
+                        if (proxy_data->username != NULL && proxy_data->password != NULL)
+                        {
+                            if(mallocAndStrcpy_s((char**)&(httpHandleData->proxy_username), (const char*)proxy_data->username) != 0)
+                            {
+                                LogError("failure allocate proxy username");
+                                free((void*)httpHandleData->proxy_host);
+                                httpHandleData->proxy_host = NULL;
+                                result = HTTPAPI_ERROR;
+                            }
+                            else
+                            {
+                                if(mallocAndStrcpy_s((char**)&(httpHandleData->proxy_password), (const char*)proxy_data->password) != 0)
```

Comment:
This leads to a possible memory leak if the option was previously set.
Check if httpHandleData->proxy_password is not NULL, then free and NULL it before calling mallocAndStrcpy_s.